include/swift/SIL/Dominance.h

```diff
@@ -184,6 +185,62 @@ class PostDominanceInfo : public PostDominatorTreeBase {
   using super::properlyDominates;
 };
 
+/// Invoke the given callback for all the reachable blocks
+/// in a function.  It will be called in a depth-first,
+/// dominance-consistent order.
+///
+/// Furthermore, prior to running each block, a tracking scope will
+/// be entered for each of the trackers passed in, as if by:
+///
+///   typename ScopedTrackingTraits<Tracker>::scope_type scope(tracker);
+///
+/// This allows state to be saved and restored for each of the trackers,
+/// such that each tracker will only represent state that was computed
+/// in a dominating block.
+template <class... Trackers, class Fn>
+void runInDominanceOrderWithScopes(DominanceInfo *dominance, Fn &&fn,
+                                   Trackers &...trackers) {
+  using TrackingStackNode = TrackingScopes<Trackers...>;
+  llvm::SmallVector<std::unique_ptr<TrackingStackNode>, 8> trackingStack;
```

Comment:
The maximum size of trackingStack is the number of blocks in the function. If you pre-allocate (e.g. reserve) an array of TrackingStackNode (not pointers!) with this size then you don't need to alloc a new StrackingStackNode for each block. Maybe just use a plain old C array.

Reply:
I can try to optimize this so that it doesn't require heap-allocation, sure.  But this is already better than CSE, FWIW.